### PR TITLE
New plugin: Write Stackdriver writes to Google Stackdriver Monitoring

### PR DESCRIPTION
**Pull request by octo**
This new plugin writes metrics to [Stackdriver Monitoring](https://cloud.google.com/monitoring/) (previously Google Cloud Monitoring).

It is similar to but not identical to Stackdriver's own *write_gcm plugin*: this implementation uses the standard `timeseries` API endpoint and *custom metrics* to write metrics. The "monitored resource" is configurable and defaults to `global`.

There is one somewhat open issue: Stackdriver expects a "start time" of "cumulative metrics" (`DERIVE`s). The first timeseries point this code is submitting has start time and end time equal, which is rejected by the API. The start time is stored in the metric's meta data and all subsequent metrics will send the same start time. However, there is a spike at the beginning of cumulative timeseries, which is suboptimal.

I realize that this PR is too big. If you want, we can treat this as the "umbrella" PR and I can split the "OAuth", "compute engine", "GCM format" and actual *write_gcm plugin* parts into separate PRs.

CC: @igorpeshansky

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -518,6 +518,68 @@ liboconfig_la_SOURCES = \
...
+libgce_la_CPPFLAGS = \
+	$(AM_CPPFLAGS) \
+	$(BUILD_WITH_LIBCURL_CFLAGS)
+libgce_la_LIBADD = \
+	$(BUILD_WITH_LIBCURL_LIBS)
+endif
+endif
+endif
+
+if BUILD_WITH_LIBYAJL2
```
Comment (rubenk) [inline Makefile.am]: This probably needs ```if BUILD_WITH_LIBSSL``` as well?

File: configure.ac [hunk 2]
```diff
@@ -5069,6 +5071,59 @@ PKG_CHECK_MODULES([LIBSIGROK], [libsigrok < 0.4],
 )
 # }}}
 
+# --with-libssl {{{
+with_libssl_cflags=""
+with_libssl_ldflags=""
+AC_ARG_WITH(libssl, [AS_HELP_STRING([--with-libssl@<:@=PREFIX@:>@], [Path to libssl.])],
```
Comment (rubenk) [inline configure.ac]: From here on, can you please keep the style of the existing configure.ac? So quote all macro arguments, if; then on the same line etc.

File: configure.ac [hunk 3]
```diff
@@ -6650,6 +6711,7 @@ AC_PLUGIN([vserver],             [$plugin_vserver],         [Linux VServer stati
 AC_PLUGIN([wireless],            [$plugin_wireless],        [Wireless statistics])
 AC_PLUGIN([write_graphite],      [yes],                     [Graphite / Carbon output plugin])
 AC_PLUGIN([write_http],          [$with_libcurl],           [HTTP output plugin])
+AC_PLUGIN([write_stackdriver],   [$plugin_write_stackdriver], [Google Stackdriver Monitoring output plugin])
```
Comment (rubenk) [inline configure.ac]: please reindent the whole AC_PLUGIN list.

File: src/daemon/utils_cache_mock.c [hunk 4]
```diff
@@ -24,9 +24,10 @@
  *   Florian octo Forster <octo at collectd.org>
  */
 
-#include <errno.h>
 #include "utils_cache.h"
 
+#include <errno.h>
```
Comment (rubenk) [inline src/daemon/utils_cache_mock.c]: unintended change?

File: src/utils_gce.c [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+  if (!curl) {
+    ERROR("utils_gce: curl_easy_init failed.");
+    return NULL;
+  }
+
+  struct curl_slist *headers = NULL;
+  headers = curl_slist_append(headers, GCE_METADATA_HEADER);
+
+  char curl_errbuf[CURL_ERROR_SIZE];
+  blob_t blob = {NULL, 0};
```
Comment (rubenk) [inline src/utils_gce.c]: `blob_t blob = {NULL}` should be enough to zero-initialize it, or am I mistaken?

File: src/utils_gce.c [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+  DEBUG("utils_gce: Checking whether I'm running on GCE ...");
+
+  CURL *curl = curl_easy_init();
+  if (!curl) {
+    ERROR("utils_gce: curl_easy_init failed.");
+    return 0;
+  }
+
+  struct curl_slist *headers = NULL;
+  headers = curl_slist_append(headers, GCE_METADATA_HEADER);
```
Comment (rubenk) [inline src/utils_gce.c]: can you combine these two lines?

File: src/utils_format_stackdriver.c [hunk 7]
```diff
@@ -0,0 +1,709 @@
...
+    return (int)status;
+
+  return 0;
+} /* }}} int json_string */
+
+static int json_time(yajl_gen gen, cdtime_t t) {
+  char buffer[64];
+  size_t status;
+
+  status = rfc3339(buffer, sizeof(buffer), t);
```
Comment (rubenk) [inline src/utils_format_stackdriver.c]: please combine this with the line above.

File: src/utils_format_stackdriver.c [hunk 8]
```diff
@@ -0,0 +1,709 @@
...
+ *   }
+ * }
+ */
+static int format_gcm_resource(yajl_gen gen, sd_resource_t *res) /* {{{ */
+{
+  int status;
+
+  yajl_gen_map_open(gen);
+
+  status = json_string(gen, "type") || json_string(gen, res->type);
```
Comment (rubenk) [inline src/utils_format_stackdriver.c]: int status

File: src/utils_format_stackdriver.c [hunk 9]
```diff
@@ -0,0 +1,709 @@
...
+
+  if (res->labels_num != 0) {
+    size_t i;
+
+    status = json_string(gen, "labels");
+    if (status != 0)
+      return status;
+
+    yajl_gen_map_open(gen);
+    for (i = 0; i < res->labels_num; i++) {
```
Comment (rubenk) [inline src/utils_format_stackdriver.c]: for (size_t i = 0...

File: src/utils_format_stackdriver.c [hunk 10]
```diff
@@ -0,0 +1,709 @@
...
+
+  unsigned char const *tmp = NULL;
+  yajl_gen_get_buf(gen, &tmp, &(size_t){0});
+  sstrncpy(buffer, (void const *)tmp, buffer_size);
+
+  yajl_gen_free(gen);
+  return 0;
+} /* }}} int sd_format_metric_descriptor */
+
+/* vim: set sw=2 sts=2 et fdm=marker : */
```
Comment (rubenk) [inline src/utils_format_stackdriver.c]: please remove any modelines from this PR, we stopped using them a while ago.

File: Makefile.am (hunk 1 above)
Comment (octo) [reply Makefile.am]: "format_stackdriver" doesn't use libssl itself, but it's exclusively used by code that does.

File: configure.ac (hunk 2 above)
Comment (octo) [reply configure.ac]: Fixed.

File: configure.ac (hunk 3 above)
Comment (octo) [reply configure.ac]: Done in 212138510373833514ed9792e9b2493bafb8ed19

File: src/daemon/utils_cache_mock.c (hunk 4 above)
Comment (octo) [reply src/daemon/utils_cache_mock.c]: More like a drive-by fix. `"utils_cache.h"` should be included first.

File: src/utils_gce.c (hunk 5 above)
Comment (octo) [reply src/utils_gce.c]: Yes. Went with `{0}` as its more commonly used to zero initialize.

**Comment by rubenk** [comment]
No worries @octo, glad to see you're back!

**Comment by vutukuribharat** [comment]
Hi 
I Need to Monitor my On-Premises Host using Google Cloud StackDriver . Kindy Suggest me the Steps.

**Pull request merged**